### DISCUSSION TITLE
Maven, What Are You Waiting For?! - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#302] [Announcements]
# Maven, What Are You Waiting For?! - Gunnar Morling

As part of my new job at Decodable,
I am also planning to contribute to the Apache Flink project
(as Decodable’s fully-managed stream processing platform is based on Flink).
Right now, I am in the process of familiarizing myself with the Flink code base,
and as such I am of course building the project from source, too.

https://www.morling.dev/blog/maven-what-are-you-waiting-for/

**Comment by cstamas** [comment]
Howdy,
While that Sonatype blog entry from 2009 was right around that time, but it is not quite true anymore (especially in time when GH Packages, and many SaaS repository manager offerings and many other things exists)... that entry was written when people ran Linux boxes in their basement "pimped" with httpd to "host" maven repositories, which today is not typical anymore...

OTOH, your entry above reminded me to point you at upcoming (3.9.0+) Maven feature, the "Remote Repository Filter", that is demoed here: https://github.com/cstamas/rrf-demo and the related resolver site https://maven.apache.org/resolver/remote-repository-filtering.html

**Reply by gunnarmorling** [reply]
Hey @cstamas, thanks for commenting and for the pointer to RRF; didn't know about that and will take a look. On the Sonatype post, what exactly is not correct about it any longer? I think it still is good advice in general to not publish artifacts with repo info to Maven Central. Also folks should publish OSS artifacts preferably on Central rather then elsewhere, sidestepping the need for custom repo info to begin with.

**Reply by cstamas** [reply]
Will try to formulate a "response post" about the topic, but in short, it is not "flat wrong", but more like "things has changed since" (and some fine print). 

Yes, ideally you do want to have all your transitive dependency hull come from Maven Central (hence no repo def in any of POM), but, that is not always doable, or just not preferred or just not always happens like it (or in other words, no artifact on Central should "point to" somewhere else, the graph of deps should lie completely on Central). 

Maven was a bit "not nice" with projects being hosted in other repositories than Maven Central, this is an oversight of the project, but things are changing...  and just to complicate things, this is somewhat in overlap with push of "repository groups" (virtual reposes) that were sadly introduced by early MRMs as a "killer feature" around same time Sonatype blog post was published, but as time passed, they proved more as problem, not a silver bullet... But again, this whole topic is not for a comment, but deserves it's own post...

**Reply by gunnarmorling** [reply]
Looking forward to reading your post, when you have the time for it.

> no artifact on Central should "point to" somewhere else, the graph of deps should lie completely on Central

Yes, that's the most important thing to me.
